From 77f0dff21086c9df1a70c4c301dfa37dcc2fcfee Mon Sep 17 00:00:00 2001
From: Michael Forney <mforney@mforney.org>
Date: Tue, 6 Jun 2017 23:30:38 -0700
Subject: [PATCH] scripts/hexdump: Support stripping prefix for symbol name

This allows us to run hexdump from a different directory, yet still have
the correct symbol names.
---
 scripts/hexdump.c | 39 +++++++++++++++++++++++++++++++--------
 1 file changed, 31 insertions(+), 8 deletions(-)

diff --git a/scripts/hexdump.c b/scripts/hexdump.c
index 728afc33..3e0ea49f 100644
--- a/scripts/hexdump.c
+++ b/scripts/hexdump.c
@@ -3,6 +3,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <unistd.h>
 
 static int
 hexdump(FILE *fo, FILE *fi)
@@ -24,26 +25,43 @@ hexdump(FILE *fo, FILE *fi)
 	return n;
 }
 
+static int
+usage(void)
+{
+	fprintf(stderr, "usage: hexdump [-p prefix] output.c input.dat\n");
+	exit(1);
+}
+
 int
 main(int argc, char **argv)
 {
 	FILE *fo;
 	FILE *fi;
-	char filename[256];
+	char filenamebuf[256], *filename = filenamebuf;
 	char *basename;
 	char *p;
-	int i, size;
+	char *prefix = "";
+	int c, i, size;
 
 	if (argc < 3)
+		usage();
+
+	while ((c = getopt(argc, argv, "0p:")) != -1)
 	{
-		fprintf(stderr, "usage: hexdump output.c input.dat\n");
-		return 1;
+		switch (c)
+		{
+		case 'p':
+			prefix = optarg;
+			break;
+		case '?':
+			usage();
+		}
 	}
 
-	fo = fopen(argv[1], "wb");
+	fo = fopen(argv[optind], "wb");
 	if (!fo)
 	{
-		fprintf(stderr, "hexdump: could not open output file '%s'\n", argv[1]);
+		fprintf(stderr, "hexdump: could not open output file '%s'\n", argv[optind]);
 		return 1;
 	}
 
@@ -55,7 +73,7 @@ main(int argc, char **argv)
 	fprintf(fo, "#endif\n");
 	fprintf(fo, "#endif\n");
 
-	for (i = 2; i < argc; i++)
+	for (i = optind + 1; i < argc; i++)
 	{
 		fi = fopen(argv[i], "rb");
 		if (!fi)
@@ -73,7 +91,7 @@ main(int argc, char **argv)
 		else
 			basename = argv[i];
 
-		if (strlen(basename) >= sizeof(filename))
+		if (strlen(basename) >= sizeof(filenamebuf))
 		{
 			fclose(fi);
 			fclose(fo);
@@ -82,6 +100,11 @@ main(int argc, char **argv)
 		}
 
 		strcpy(filename, argv[i]);
+		while (*prefix && *prefix == *filename)
+		{
+			++prefix;
+			++filename;
+		}
 		for (p = filename; *p; ++p)
 		{
 			if (*p == '/' || *p == '.' || *p == '\\' || *p == '-')
-- 
2.13.1

